File: cpp/src/preprocess.cc
#include "preprocess.h"

cv::Mat centerCrop(const cv::Mat& img, int targetWidth, int targetHeight) {
    int height = img.rows;
    int width = img.cols;
    int startX = width / 2 - targetWidth / 2;
    int startY = height / 2 - targetHeight / 2;
    return img(cv::Rect(startX, startY, targetWidth, targetHeight));
}

cv::Mat eulerToRotationMatrix(float roll, float pitch, float yaw) {
    cv::Mat R_x = (cv::Mat_<double>(3, 3) <<
        1, 0, 0,
        0, cos(roll), -sin(roll),
        0, sin(roll), cos(roll));

    cv::Mat R_y = (cv::Mat_<double>(3, 3) <<
        cos(pitch), 0, sin(pitch),
        0, 1, 0,
        -sin(pitch), 0, cos(pitch));

    cv::Mat R_z = (cv::Mat_<double>(3, 3) <<
        cos(yaw), -sin(yaw), 0,
        sin(yaw), cos(yaw), 0,
        0, 0, 1);
    cv::Mat R = R_z * R_y * R_x;
    return R;
}

cv::Mat distort(const std::string& pathStr) {

    // std::string pathStr(imgPath);
    cv::Mat img = cv::imread(pathStr);
    if (img.empty()) {
        printf("Could not open or find the image!\n");
        return cv::Mat();
    }

    int height = img.rows;
    int width = img.cols;
    printf("image height: %d, width: %d\n", height, width);

    // Extract roll, pitch, yaw from the filename
    // Assuming the filename format is something like "image@roll@pitch@yaw.jpg"
    size_t atPos = pathStr.find('@');
    atPos = pathStr.find('@', atPos + 1);
    atPos = pathStr.find('@', atPos + 1);

    float yaw = std::stof(pathStr.substr(atPos + 1));
    atPos = pathStr.find('@', atPos + 1);
    float pitch = std::stof(pathStr.substr(atPos + 1));
    atPos = pathStr.find('@', atPos + 1);
    float roll = std::stof(pathStr.substr(atPos + 1));

    printf("yaw: %f, pitch: %f, roll: %f\n", yaw, pitch, roll);
    

    cv::Mat R = eulerToRotationMatrix(roll, pitch, yaw);
    // Define the source points (four corners of the image)
    std::vector<cv::Point2f> srcPoints = {
        cv::Point2f(0, 0),
        cv::Point2f(width - 1, 0),
        cv::Point2f(width - 1, height - 1),
        cv::Point2f(0, height - 1)
    };
    // Apply the rotation to the source points
    std::vector<cv::Point2f> dstPoints(4);
    for (size_t i = 0; i < srcPoints.size(); ++i) {
        cv::Mat pt = (cv::Mat_<double>(2, 1) << srcPoints[i].x - width / 2.0, srcPoints[i].y - height / 2.0);
        cv::Mat dstPt = R(cv::Rect(0, 0, 2, 2)) * pt;
        dstPoints[i] = cv::Point2f(dstPt.at<double>(0) + width / 2.0, dstPt.at<double>(1) + height / 2.0);
    }
    // Compute the perspective transform matrix
    cv::Mat matrix = cv::getPerspectiveTransform(srcPoints, dstPoints);

    // Apply the perspective transformation
    cv::Mat adjustedImage;
    cv::warpPerspective(img, adjustedImage, matrix, img.size());
    return adjustedImage;
}

cv::Mat preProcess(const std::string& pathStr, float contrastFactor) {
    // Load and distort image
    cv::Mat img = distort(pathStr);


    // Center crop
    img = centerCrop(img);

    // Convert to grayscale
    cv::Mat gray;
    cv::cvtColor(img, gray, cv::COLOR_BGR2GRAY);

    // Adjust contrast using contrastFactor
    gray.convertTo(gray, CV_32F, 1 / 255.0);
    float mean = cv::mean(gray)[0];
    gray = mean + contrastFactor * (gray - mean);

    // Clip values to [0, 1]
    cv::threshold(gray, gray, 1, 1, cv::THRESH_TRUNC);
    cv::threshold(gray, gray, 0, 0, cv::THRESH_TOZERO);

    // Convert grayscale image to 3-channel image
    // 将灰度图像复制到三个通道
    cv::Mat image;
    cv::merge(std::vector<cv::Mat>{gray, gray, gray}, image);

    // 转换图像为浮点类型
    image.convertTo(image, CV_32F);

    // 均值和标准差
    static const float meanVals[3] = {0.485, 0.456, 0.406};
    static const float stdVals[3] = {0.229, 0.224, 0.225};

    // 拆分图像为独立通道
    std::vector<cv::Mat> channels(3);
    cv::split(image, channels);

    // 归一化每个通道
    for (int i = 0; i < 3; ++i) {
        channels[i] = (channels[i] - meanVals[i]) / stdVals[i];
    }

    // 合并归一化后的通道
    cv::merge(channels, image);

    return image;
}


// int main() {
//     std::string imgPath = "/home/xujg/code/RK_demo/rknn_model_zoo-main/examples/uav_vision_loc/model/@437915@4221754@1.484649@-0.183243@0.004551@203.8990020752@39093@.jpg";
//     cv::Mat processedImg = preProcess(imgPath);
//     if (!processedImg.empty()) {
//         cv::imwrite("./demo_output.jpg", processedImg);
//     }
//     return 0;
// }

File: python/py_utils/pre_process.py
import numpy as np
import cv2
from .logger_config import logger
import sys


def center_crop(img, target_width=512, target_height=512):
    height, width, _ = img.shape
    start_x = width // 2 - target_width // 2
    start_y = height // 2 - target_height // 2
    return img[start_y:start_y + target_height, start_x:start_x + target_width]

def euler_to_rotation_matrix(yaw, pitch, roll):
    R_x = np.array([[1, 0, 0], [0, np.cos(roll), -np.sin(roll)], [0, np.sin(roll), np.cos(roll)]])
    R_y = np.array([[np.cos(pitch), 0, np.sin(pitch)], [0, 1, 0], [-np.sin(pitch), 0, np.cos(pitch)]])
    R_z = np.array([[np.cos(yaw), -np.sin(yaw), 0], [np.sin(yaw), np.cos(yaw), 0], [0, 0, 1]])
    return np.dot(R_z, np.dot(R_y, R_x))

def distort(img, at):
    height, width = img.shape[:2]
    R = euler_to_rotation_matrix(at[0], at[1], at[2])
    src_points = np.array([[0, 0], [width - 1, 0], [width - 1, height - 1], [0, height - 1]], dtype='float32')
    dst_points = np.dot(src_points - np.array([width / 2, height / 2]), R[:2, :2].T) + np.array([width / 2, height / 2])
    matrix = cv2.getPerspectiveTransform(src_points, dst_points.astype('float32'))
    return cv2.warpPerspective(img, matrix, (width, height))

def pre_process(data, at=[0., 0., 0.]):

    if isinstance(data, np.ndarray):
        img = data
    elif isinstance(data, str):
        img = cv2.imread(data)
        at = list(map(np.float32, data.split("@")[3:6]))
    at_info = (
        f"yaw: {at[0]}, "
        f"pitch: {at[1]}, "
        f"roll: {at[2]}, "
    )
    logger.debug(at_info)
    img = distort(img, at)
    img = center_crop(img)
    # if img.shape[0] != 512 or img.shape[1] != 512:
    #     logger.error("input size error.")
    img_gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    img_float = img_gray.astype(np.float32) / 255.0
    mean = np.mean(img_float)
    img_contrast = mean + 3 * (img_float - mean)
    img_clipped = np.clip(img_contrast, 0, 1)
    img_rgb = np.repeat(img_clipped[:, :, np.newaxis], 3, axis=2)
    mean = np.array([0.485, 0.456, 0.406])
    std = np.array([0.229, 0.224, 0.225])
    img_normalized = (img_rgb - mean) / std
    return img_normalized.astype(np.float32)



